Fix ambient cni improper retry

fyuan1316 opened the pull request:
**Please provide a description of this PR:**



**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**
close #49875 
- [x] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

istio-testing commented:
Hi @fyuan1316. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

zengyuxing007 commented:
/ok-to-test

zengyuxing007 reviewed (approved):
LGTM

istio-testing commented:
@fyuan1316: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 59c0e97bd8d5583b5e2918c586c64d1f0922e2a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/release-notes_istio/1767897926534696960) | true | `/test release-notes`
integ-pilot_istio | 59c0e97bd8d5583b5e2918c586c64d1f0922e2a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/integ-pilot_istio/1767897918984949760) | true | `/test integ-pilot`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented on cni/pkg/nodeagent/cni-watcher.go:
```diff
@@ -169,8 +169,8 @@ func (s *CniPluginServer) ReconcileCNIAddEvent(ctx context.Context, addCmd CNIPl
 	// The plugin already consulted the k8s API - but on this end handler caches may be stale, so retry a few times if we get no pod.
-	for ambientPod, err = s.handlers.GetPodIfAmbient(addCmd.PodName, addCmd.PodNamespace); (ambientPod == nil) && (retries < maxStaleRetries); retries++ {
-		if err != nil {
+	for ; (ambientPod == nil) && (retries <= maxStaleRetries); retries++ {
```
There is also `pkg/backoff` which could probably be used

fyuan1316 replied:
@bleggett 
It's great to see you helping out with the review .I've recently been learning about ambient in-pod implementations, and the discovery of this little problem came from that learning process. This novel design is very interesting, thank you very much for your awesome work. 
Sorry for submitting a flawed version of the fix. Yes, I agree that increasing test case coverage always helps.

`There is also pkg/backoff which could probably be used` 
@howardjohn good suggestion!

bleggett replied:
@fyuan1316 I appreciate you taking a look at it!

istio-testing commented:
@fyuan1316: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot_istio | 59c0e97bd8d5583b5e2918c586c64d1f0922e2a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/integ-pilot_istio/1767897918984949760) | true | `/test integ-pilot`
release-notes_istio | 767c05de84f189ae23a636ddf7403f213d18e76b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/release-notes_istio/1768152341711687680) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@fyuan1316: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot_istio | 59c0e97bd8d5583b5e2918c586c64d1f0922e2a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/integ-pilot_istio/1767897918984949760) | true | `/test integ-pilot`
release-notes_istio | 767c05de84f189ae23a636ddf7403f213d18e76b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/release-notes_istio/1768152341711687680) | true | `/test release-notes`
integ-cni_istio | 767c05de84f189ae23a636ddf7403f213d18e76b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/integ-cni_istio/1768152330265432064) | true | `/test integ-cni`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

fyuan1316 commented:
@bleggett 
Fixed always warn issue and added unit test for retry logic.

The reason we didn't use `pkg/backoff` here is that the method `RetryWithContext()` stops when no error is generated, and we want to get and return possible errors for the method `s.handlers.GetPodIfAmbient(...)` . This makes error handling a bit inconvenient.  @howardjohn 

Do you guys think this is ok?

istio-testing commented:
@fyuan1316: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-cni_istio | 767c05de84f189ae23a636ddf7403f213d18e76b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/integ-cni_istio/1768152330265432064) | true | `/test integ-cni`
release-notes_istio | 642153fd370f4f6fed3ec1d5c9c613b9711e166d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/release-notes_istio/1768187755642228736) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

bleggett commented on cni/pkg/nodeagent/cni-watcher_test.go:
```diff
@@ -216,3 +217,26 @@ func TestCNIPluginServerPrefersCNIProvidedPodIP(t *testing.T) {
+	meshDataplane := &meshDataplane{kubeClient: client.Kube()}
+	pluginServer := startCniPluginServer(ctx, "/tmp/test.sock", retryableHandler, meshDataplane)
+	_ = pluginServer.ReconcileCNIAddEvent(ctx, CNIPluginAddEvent{})
+	require.Greater(t, len(retryableHandler.counter), 1, "GetPodIfAmbient should not be executed only once due to the retry policy.")
```
I like/prefer the `testify` assert libs too, but this `require` can be expressed with a plain `assert` from `"istio.io/istio/pkg/test/util/assert"` which is already imported, so we don't need to add a new dep.

Also, this would retry the max amount of times, correct? This is synchronous so we will always retry 10x, so I don't think anything is gained by checking "at least one"

fyuan1316 replied:
>Also, this would retry the max amount of times, correct? This is synchronous so we will always retry 10x, so I don't think anything is gained by checking "at least one"

Since the loop terminates with `(ambientPod == nil) && (retries <= maxStaleRetries)`. So I don't think there will be a case where it always executes 10 times. Also I think the elapsed time here should be around [0 ~ 10x10ms]. 0 is the first time a valid ambientPod is queried, 10x10ms is the retries until all the chances are consumed.

By "synchronization" do you mean the "counter channel" in "retryableHandler"? If so, since it's a 20 size buffer channel here, I don't think we need to worry about that, there won't be a synchronization block problem.

@bleggett 
If you want, I can also add cases to the unit test for the two scenarios _error occurred_ and _successful fetch_, maybe that might be better?

Thank you for your time.

fyuan1316 replied:
Sorry, it dawned on me that you were probably talking about retrying the test 10 times, which is true in unit tests.

bleggett commented:
> @bleggett Fixed always warn issue and added unit test for retry logic.
> 
> The reason we didn't use `pkg/backoff` here is that the method `RetryWithContext()` stops when no error is generated, and we want to get and return possible errors for the method `s.handlers.GetPodIfAmbient(...)` . This makes error handling a bit inconvenient. @howardjohn
> 
> Do you guys think this is ok?

If you want, you can move the retry logic inside `s.handlers.GetPodIfAmbient(...)` itself, which should let you use `backoff` more easily.

istio-testing commented:
@fyuan1316: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 52e206e6f5472b8019e05becd1678e515d72a6da | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49876/release-notes_istio/1768484645101375488) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

fyuan1316 commented:
>I like/prefer the testify assert libs too, but this require can be expressed with a plain assert from "istio.io/istio/pkg/test/util/assert" which is already imported, so we don't need to add a new dep.

Use assert instead of require. @bleggett 

I'm not sure that we can tolerate unit tests with a fixed 100 millisecond elapsed time. Personally, if this is a big deal, I'd probably suggest removing unit tests. Personally, instead of adding retry logic to the `s.handlers.GetPodIfAmbient` method, I think it would be better to keep its responsibilities simple.

bleggett commented on cni/pkg/nodeagent/cni-watcher_test.go:
```diff
@@ -216,3 +216,26 @@ func TestCNIPluginServerPrefersCNIProvidedPodIP(t *testing.T) {
+	meshDataplane := &meshDataplane{kubeClient: client.Kube()}
+	pluginServer := startCniPluginServer(ctx, "/tmp/test.sock", retryableHandler, meshDataplane)
+	_ = pluginServer.ReconcileCNIAddEvent(ctx, CNIPluginAddEvent{})
+	assert.Equal(t, len(retryableHandler.counter) > 1, true, "GetPodIfAmbient should not be executed only once due to the retry policy.")
```
```suggestion
	assert.Equal(t, len(retryableHandler.counter) , 10, "GetPodIfAmbient should not be executed only once due to the retry policy.")
```
What I was saying before was since in the test it will always exec 10 times no matter what there's no need to have a fuzzy check for "> 1"

bleggett commented:
> > I like/prefer the testify assert libs too, but this require can be expressed with a plain assert from "istio.io/istio/pkg/test/util/assert" which is already imported, so we don't need to add a new dep.
> 
> Use assert instead of require. @bleggett
> 
> I'm not sure that we can tolerate unit tests with a fixed 100 millisecond elapsed time. Personally, if this is a big deal, I'd probably suggest removing unit tests. Personally, instead of adding retry logic to the `s.handlers.GetPodIfAmbient` method, I think it would be better to keep its responsibilities simple.

Eh, if it makes behavior easier to test I don't mind, it's not a big increase in responsibilities - it's only called here, it's specific to the ambient informer so it will only be used in contexts where cache misses are possible, and you can extend the method signature with simple retry opts so the unit test can have a smaller retry threshold.
